source/AI.cpp

```diff
@@ -2066,6 +1898,194 @@ bool AI::ShouldDock(const Ship &ship, const Ship &parent, const System *playerSy
+			-> shared_ptr<Ship>
+		{
+			for(const auto &other : otherShips)
+				if(other->GetGovernment() == gov && other->GetSystem() == system && !other->CanBeCarried())
+				{
+					if(!other->IsDisabled() && other->CanCarry(*ship))
+						return other;
+					else
+						parentChoices.emplace_back(other);
+				}
+			return shared_ptr<Ship>();
+		};
+		// Mission ships should only pick amongst ships from the same mission.
+		auto missionIt = ship->IsSpecial() && !isYours
+			? find_if(player.Missions().begin(), player.Missions().end(),
+				[&ship](const Mission &m) { return m.HasShip(ship); })
+			: player.Missions().end();
+
+		shared_ptr<Ship> newParent;
+		if(missionIt != player.Missions().end())
+		{
+			const auto &npcs = missionIt->NPCs();
+			for(const auto &npc : npcs)
+			{
+				// Don't reparent to NPC ships that have not been spawned.
+				if(!npc.ShouldSpawn())
+					continue;
+				newParent = getParentFrom(npc.Ships());
+				if(newParent)
+					break;
+			}
+		}
+		else
+			newParent = getParentFrom(ships);
+
+		// If a new parent was found, then this carried ship should always reparent
+		// as a ship of its own government is in-system and has space to carry it.
+		if(newParent)
+			parent = newParent;
+		// Otherwise, if one or more in-system ships of the same government were found,
+		// this carried ship should flock with one of them, even if they can't carry it.
+		else if(!parentChoices.empty())
+			parent = parentChoices[Random::Int(parentChoices.size())];
+		// Player-owned carriables that can't be carried and have no ships to flock with
+		// should keep their current parent, or if it is destroyed, their parent's parent.
+		else if(isYours)
+		{
+			if(parent && parent->IsDestroyed())
+				parent = parent->GetParent();
+		}
+		// All remaining non-player ships should forget their previous parent entirely.
+		else
+			parent.reset();
+
+		// Player-owned carriables should defer to player carrier if
+		// selected parent can't carry it. This is necessary to prevent
+		// fighters from jumping around fleet when there's not enough
+		// bays.
+		if(isYours && parent && parent->GetParent() && !parent->CanCarry(*ship))
+			parent = parent->GetParent();
+
+		if(ship->GetParent() != parent)
+			ship->SetParent(parent);
+	}
+	// Otherwise, check if this ship wants to return to its parent (e.g. to repair).
+	else if(parentHasSpace && ShouldDock(*ship, *parent, playerSystem))
+	{
+		ship->SetTargetShip(parent);
+		MoveTo(*ship, command, parent->Position(), parent->Velocity(), 40., .8);
+		command |= Command::BOARD;
+		ship->SetCommands(command, firingCommands);
+		return true;
+	}
+	// If we get here, it means that the ship has not decided to return
+	// to its mothership. So, it should continue to be deployed.
+	command |= Command::DEPLOY;
+
+	return false;
+}
+
+
+
+bool AI::HandlePersonalityTraits(const shared_ptr<Ship> &ship, shared_ptr<Ship> &parent, Command &command,
```

Comment:
I'm not too sure about this function since some personality things are still done in AI::Step.